### DISCUSSION TITLE
AWS lamda@edge를 활용한 이미지 리사이징 자동화

### DISCUSSION THREAD

**Discussion by lina1919** [SWM-Space-Odyssey/WeddingMate_BackEnd#29] [General]
## AWS lamda@edge 서버 아키텍처
<img width="782" alt="image" src="https://github.com/SWM-Space-Odyssey/WeddingMate_BackEnd/assets/63230463/6eaed3f2-d837-48de-8225-45e768dc63f4">

- on the fly 아키텍처 
[https://mystworld.tistory.com/38](url)
- edge 컴퓨팅 : 첫요청에만 처리, 그 다음 사용자 요청부터 caching
- width, height, quality, 이미지 포멧 변환의 리사이징 가능

### 기타활용
- 보안등급 올릴 때도 사용
- 검색 엔진 최적화(SEO) -> AB testing, 크롤링이나 디도스 방지 등에도 활용 가능
- signed url 기능 : 특정 Cloud Storage 리소스에 대한 제한된 시간 동안 액세스 권한을 제공하는 URL(cloudfront_signer)

### WeddingMate에 도입 가능성
- 이미지 서버는 serverless로 구축하여 빠른 이미지 변환 및 리사이징 가능
- 이미지 해상도 처리를 따로 백엔드에서 해주지 않아도 됨
- S3와 연결하여 바로 사용할 수 있음
- 다만 @CSKIM999 은 로딩 시간이 S3에서 바로 불러올 때보다 오래 걸린다고 느낌

[https://d2wbg53x3mvi56.cloudfront.net/nature-3039901_1280.jpg?w=700](url)
- 아래 링크에서 요청 처리해보시고 의견 주면 감사할 것 같습니다!

**Comment by nitronium102** [comment]
처음에 로딩할 때는 조금 느린 것 같은데, 그래도 여러 이점들을 생각해봤을 때 도입하면 좋을 것 같습니다!

**Comment by lina1919** [comment]
한글 url로 된 s3 이미지 파일을 불러오는데 오류가 나는 것을 발견하였습니다.(By. @CSKIM999)
그 이유는 lambda에서 url을 읽을 때 자동으로 인코딩하여 읽어오기 때문입니다.
따라서 URL 인코딩된 문자열을 다시 원래의 문자열로 해독하는 코드를 javascript에서 기본적으로 제공하는 decodeURIComponent 함수를 사용하여 추가하여 해결하였습니다.
